app/src/main/java/mozilla/lockbox/store/NetworkStore.kt

```diff
@@ -0,0 +1,68 @@
+package mozilla.lockbox.store
+
+import android.content.Context
+import android.net.ConnectivityManager
+import io.reactivex.Observable
+import io.reactivex.rxkotlin.addTo
+import mozilla.lockbox.action.NetworkAction
+import mozilla.lockbox.extensions.filterByType
+import mozilla.lockbox.flux.Dispatcher
+import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.subjects.ReplaySubject
+
+open class NetworkStore(
+    val dispatcher: Dispatcher = Dispatcher.shared
+) : ContextStore {
+
+    open lateinit var connectivityManager: ConnectivityManager
+    internal val compositeDisposable = CompositeDisposable()
+
+    private val stateSubject: ReplaySubject<Boolean> = ReplaySubject.createWithSize(1)
+    open val networkAvailable: Observable<Boolean> get() = stateSubject
+
+    // MAKE ME PRIVATE
+    private val networkConnectivityState: Boolean
+        get() = connectivityManager.activeNetworkInfo?.isConnectedOrConnecting == true
+
+    companion object {
+        val shared = NetworkStore()
+    }
+
+//    sealed class State {
+//        val Connected : Boolean()
+//
+//        object ConnectionError : Boolean = false
+//        data class ConnectionError(val error: String? = null) : State()
+//    }
+
+    init {
+        dispatcher.register
+            .filterByType(NetworkAction::class.java)
+            .subscribe {
+                when (it) {
+                    is NetworkAction.CheckConnectivity -> {
+                        checkConnectivity()
+                    }
+                }
+            }
+            .addTo(compositeDisposable)
+
+        stateSubject.subscribe { state ->
+                networkAvailable.doOnNext { state }
+        }.addTo(compositeDisposable)
+    }
+
+    override fun injectContext(context: Context) {
+        connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+    }
+
+    private fun checkConnectivity() {
+        if (networkConnectivityState){
+
+            stateSubject.onNext(true)
+        }
+        else {
+            stateSubject.onNext(false)
+        }
+    }
+}
```

Comment:
I'm not sure I like this approach, here. 

We're only checking the network when you send a `CheckConnectivity` action, and then immediately getting a response back; I don't think it's very much different from simply calling the method synchronously and directly.

I _think_ there is functionality in the `ConnectivityManager` to add a listener to listen to changes in the network connection. The listener then feeds the `stateSubject`.

e.g. [`registerDefaultNetworkCallback`][1], with a fallback to a BroadcastReceiver for the right API levels.

[1]: https://developer.android.com/reference/android/net/ConnectivityManager.html#registerDefaultNetworkCallback(android.net.ConnectivityManager.NetworkCallback)

Once you get a working network connection, there are few interesting things to do here:

1. Throttle the `networkAvailability` observable. Spotty network connections cause a lot of changes of network connectivity. Would be great if we knew that the network has been up for at least a few seconds.
2. Check the liveness of the connection. Pinging a well known URL when you have a connection to check that you can actually seen the internet (rather than just the proximal network link) before you update the `networkAvailability` to `true`.

This store would be a good candidate to turn into an Android Component. Could you also file a story for that, please?

Reply:
Todo

Reply:
https://github.com/mozilla-lockbox/lockbox-android/issues/355

Reply:
There are two parts to the above comment, only the second is captured by #355.

We should strive to get a nicer API, and build a sub-optimal implementation so that when we build a better implementation there is less code to change and test.

I think we should make a call to check network connectivity on `doOnSubscribe()`,  so that we don't have to manually call `dispatcher.dispatch(CheckConnectivity)`.

Reply:
Added #369 for the longer-term A-C stuff, and changed #355 to reflect that first part about adding a listener to the network connectivity manager